sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/HashUtils.scala

```diff
@@ -0,0 +1,135 @@
+    private val sb = new StringBuilder()
+    def result: String = sb.result()
+    override def write(b: Int): Unit =
+      discard(sb.append(Bytes.fromByteArray(Array(b.toByte)).toHexString))
+    override def write(b: Array[Byte], off: Int, len: Int) = {
+      discard(sb.append(Bytes.fromByteArray(b.slice(off, off + len)).toHexString))
+    }
+    override def write(byte: Byte): Unit =
+      discard(sb.append(formatByteToHexString(byte)))
+
+    def writeContext(context: String): Unit = discard(sb.append(context))
+
+    override def close(): Unit = sb.clear()
+  }
+
+  /** Interface for message digest. Allows to swap out default implementation for debug implementation.
+    */
+  private[crypto] trait MessageDigestWithContext {
+    def messageDigest: MessageDigest
+    def update(a: Byte, context: Option[String]): Unit
+    def update(a: ByteBuffer, context: Option[String]): Unit
+    def update(a: Array[Byte], context: Option[String]): Unit
+    def digest(buf: Array[Byte], offset: Int, len: Int): Int
+    def addContext(context: String): Unit
+    def isContextAware: Boolean
+  }
+
+  /** Default Message Digest implementation. Ignores context and simply feeds all bytes to the delegate MessageDigest.
+    * @param messageDigest MessageDigest delegate which calculates the hash.
+    */
+  private[crypto] class DefaultMessageDigest(override val messageDigest: MessageDigest)
+      extends MessageDigestWithContext {
+    override val isContextAware: Boolean = false
+    override def update(a: Byte, context: Option[String]): Unit = messageDigest.update(a)
+    override def update(a: ByteBuffer, context: Option[String]): Unit = messageDigest.update(a)
+    override def update(a: Array[Byte], context: Option[String]): Unit = messageDigest.update(a)
+    override def digest(buf: Array[Byte], offset: Int, len: Int): Int =
+      messageDigest.digest(buf, 0, underlyingHashLength)
+    // Default message digest doesn't do anything with the context
+    override def addContext(context: String): Unit = {}
+  }
+
+  /** Message Digest for debugging with support for contextual information and output stream to observe the encoding result.
+    *
+    * @param messageDigest MessageDigest delegate which calculates the hash.
+    * @param outputStream Output stream that also receives the encoded objects along with context information.
+    *
+    * The format of the produced debug output is the following:
+    *
+    * # Context About Following Lines
+    * "encoded value in hex" - ["original value before encoding" ("additional contextual info", for instance type of the value)]
+    *
+    * e.g:
+    *   # Package Name
+    *   0000000e - [14 (int)]
+    *   7061636b6167652d6e616d652d30 - [package-name-0 (string)]
+    */
+  private[crypto] class DebugMessageDigest(
+      messageDigest: MessageDigest,
+      outputStream: ContextAwareOutputStream,
+  ) extends DefaultMessageDigest(messageDigest) {
```

Comment:
I had rather not `DebugMessageDigest` be a subtype of `DefaultMessageDigest` because then we cannot express statically that something is not a `DebugMessageDigest`.